File: csharp/src/Customer.cs
using System;
using System.Collections.Generic;

namespace RefactoringKata
{
    public class Customer
    {
        private string name;
        private List<Rental> rentals;

        public Customer(string name)
        {
            this.name = name;
            this.rentals = new List<Rental>();
        }

        public void AddRental(Rental rental)
        {
            rentals.Add(rental);
        }

        public string GetName()
        {
            return name;
        }

        public string Statement()
        {
            double totalAmount = 0.0;
            int frequentRenterPoints = 0;
            string result = "Rental Record for " + GetName() + "\n";

            foreach (var rental in rentals)
            {
                double thisAmount = 0.0;
                switch (rental.GetMovie().GetPriceCode())
                {
                    case Movie.CHILDRENS:
                        thisAmount += 1.5;
                        if (rental.GetDaysRented() > 3)
                            thisAmount += (rental.GetDaysRented() - 3) * 1.5;
                        break;
                    case Movie.REGULAR:
                        thisAmount += 2.0;
                        if (rental.GetDaysRented() > 2)
                            thisAmount += (rental.GetDaysRented() - 2) * 1.5;
                        break;
                    case Movie.NEW_RELEASE:
                        thisAmount += rental.GetDaysRented() * 3.0;
                        break;
                }
                frequentRenterPoints++;
                if (rental.GetMovie().GetPriceCode() == Movie.NEW_RELEASE && rental.GetDaysRented() > 1)
                {
                    frequentRenterPoints++;
                }
                result += "\t" + rental.GetMovie().GetTitle() + "\t" + thisAmount + "\n";
                totalAmount += thisAmount;
            }
            result += "You owed " + totalAmount + "\n";
            result += "You earned " + frequentRenterPoints + " frequent renter points\n";
            return result;
        }
    }
}

File: typescript/src/customer.ts
import Movie from './movie';
import Rental from './rental';

class Customer {
  name: String;
  rentals: Rental[];

  constructor(name: String) {
    this.name = name;
    this.rentals = [];
  }

  addRental(rental: Rental) {
    this.rentals.push(rental);
  }

  getName(): String {
    return this.name;
  }

  statement(): String {
    var totalAmount = 0.0;
    var frequentRenterPoints = 0;
    var result = "Rental Record for " + this.getName() + "\n";

    this.rentals.forEach(it => {
      var thisAmount = 0.0;

      switch (it.getMovie().getPriceCode()) {
        case Movie.CHILDRENS():
          thisAmount += 1.5;
          if (it.getDaysRented() > 3)
              thisAmount += (it.getDaysRented() - 3) * 1.5;
          break;
        case Movie.REGULAR():
          thisAmount += 2.0;
          if (it.getDaysRented() > 2) {
            thisAmount += (it.getDaysRented() - 2) * 1.5;
          }
          break;
        case Movie.NEW_RELEASE():
          thisAmount += it.getDaysRented() * 3;
          break;
      }

      frequentRenterPoints++;
      if (it.getMovie().getPriceCode() == Movie.NEW_RELEASE() && it.getDaysRented() > 1) {
          frequentRenterPoints++;
      }
      
      result += "\t" + it.getMovie().getTitle() + "\t" + thisAmount + "\n";
      totalAmount += thisAmount;
    })

    result += "You owed " + totalAmount + "\n";
    result += "You earned " + frequentRenterPoints + " frequent renter points\n";

    return result;
  }

}

export default Customer;